client/src/main/java/org/asynchttpclient/util/AuthenticatorUtils.java

```diff
@@ -81,25 +89,307 @@ private static String computeDigestAuthentication(Realm realm, Uri uri) {
+            try {
+                md.update(request.getByteData());
+                return MessageDigestUtils.bytesToHex(md.digest());
+            } finally {
+                md.reset();
+            }
+        }
+
+        // Handle BodyGenerator
+        if (request.getBodyGenerator() != null) {
+            return bufferAndHashBodyGenerator(request.getBodyGenerator(), hashAlgorithm);
+        }
+
+        throw new IllegalStateException("Unexpected request body state");
 
-        // FIXME isn't there a more efficient way?
-        return new String(StringUtils.charSequence2Bytes(builder, ISO_8859_1), StandardCharsets.UTF_8);
     }
 
+    /**
+     * Resolve the charset used to read / hash a request body.
+     * Order of precedence:
+     * 1) request.getCharset()      – per-request override
+     * 2) realm.getCharset()        – negotiated via RFC 7616 (e.g. UTF-8)
+     * 3) ISO-8859-1                – RFC default
+     */
+    private static Charset resolveCharset(Request request, Realm realm) {
+        Charset cs = request.getCharset();
+        if (cs != null) {
+            return cs;
+        }
+        cs = realm.getCharset();
+        return (cs != null) ? cs : StandardCharsets.ISO_8859_1;
+    }
+
+    /**
+     * Buffers the body from the given BodyGenerator and computes its hash.
+     * This is used for auth-int where the body needs to be hashed.
+     *
+     * @param gen           The BodyGenerator to read from
+     * @param hashAlgorithm The hash algorithm to use (e.g., "MD5", "SHA-256")
+     * @return The hex string of the computed hash
+     * @throws UnsupportedOperationException if the body is too large or unsupported type
+     */
+    private static String bufferAndHashBodyGenerator(BodyGenerator gen, String hashAlgorithm) {
+        MessageDigest md = MessageDigestUtils.pooledMessageDigest(hashAlgorithm);
+        // Size guard
+        if (gen instanceof ByteArrayBodyGenerator) {
+            ByteArrayBodyGenerator bag = (ByteArrayBodyGenerator) gen;
+
+            long size = bag.createBody().getContentLength();
+            if (size > MAX_AUTH_INT_BODY_SIZE) {
+                throw new UnsupportedOperationException("auth-int not supported for ByteArrayBodyGenerator >10 MB");
+            }
+        } else if (gen instanceof FileBodyGenerator) {
+            FileBodyGenerator fg = (FileBodyGenerator) gen;
+
+            long fileSize = fg.getFile().length();
+            if (fileSize > MAX_AUTH_INT_BODY_SIZE) {
+                throw new UnsupportedOperationException("auth-int not supported for files > 10 MB");
+            }
+            try {
+                byte[] bytes = Files.readAllBytes(fg.getFile().toPath());  // may throw IOException
+                md.update(bytes);
+                return MessageDigestUtils.bytesToHex(md.digest());
+            } catch (IOException ioe) {
+                throw new RuntimeException("Failed to read file for auth-int hash", ioe);
```

Comment:
This should be IOException not RuntimeException, because it is caused by an IO operation.